File: mobile/android/app/src/main/kotlin/sh/openagi/mobile/widget/WidgetState.kt
package sh.openagi.mobile.widget

import sh.openagi.mobile.protocol.Counts
import sh.openagi.mobile.protocol.TaskItem
import sh.openagi.mobile.store.Snapshot
import java.time.Instant

sealed class WidgetState {
    object Unpaired : WidgetState()
    data class Empty(val headline: String) : WidgetState()
    data class Tasks(val items: List<TaskItem>, val counts: Counts, val ageMinutes: Int) : WidgetState()
    data class Stale(val ageMinutes: Int) : WidgetState()

    // The last refresh attempt did not reach the daemon. Distinct from Stale:
    // Stale is purely an age judgement ("we haven't heard in a while, we don't
    // know why"); Unreachable is a fact we actually have ("we just tried and
    // failed"), which can be true well inside the 60-minute staleness window.
    // Without this case a daemon that has been down the whole time renders
    // identically to a healthy one until the age crosses STALE_AFTER_MINUTES.
    // Still shows the last-known tasks — DESIGN.md: "every screen renders from
    // the last snapshot when the daemon is unreachable" — just labelled
    // honestly instead of claiming the row list is current.
    data class Unreachable(val items: List<TaskItem>, val counts: Counts, val ageMinutes: Int) : WidgetState()

    companion object {
        // Past this, old rows stop being information and start being a lie.
        const val STALE_AFTER_MINUTES = 60

        fun from(snapshot: Snapshot?, paired: Boolean, now: Instant = Instant.now()): WidgetState {
            if (!paired) return Unpaired
            if (snapshot == null) return Empty("Open OpenAGI to sync")
            val age = snapshot.ageInMinutes(now)
            if (snapshot.lastRefreshFailed) return Unreachable(snapshot.visibleToday, snapshot.visibleCounts, age)
            if (age > STALE_AFTER_MINUTES) return Stale(age)
            val visible = snapshot.visibleToday
            if (visible.isEmpty()) return Empty(snapshot.summary.brief.headline.ifEmpty { "Nothing due today" })
            return Tasks(visible, snapshot.visibleCounts, age)
        }
    }
}

File: mobile/ios/Sources/Store/WidgetState.swift
import Foundation
import WidgetKit

// The identifier both processes must agree on: WidgetCenter matches a
// reload request to a widget by exactly this string. Defined once, here in
// Sources/ (not inside the `Widget`-conforming `TodayWidget`, which is
// implicitly @MainActor-isolated and so cannot be read from the AppIntent's
// non-isolated `perform()`), so the widget's own declaration and every
// caller that reloads it -- the app's RefreshCoordinator and this
// extension's own CompleteTaskIntent -- cannot drift apart.
public enum TodayWidgetKind {
    public static let value = "TodayWidget"
}

// The one state machine both the iOS and Android widgets render from. Pure
// function on purpose: it takes `paired` as a parameter instead of reading
// the Keychain itself, because a unit-test bundle cannot arrange Keychain
// state, and because this keeps it the literal twin of Android's
// `WidgetState.from`. The timeline provider is what actually supplies
// `paired: Credentials.load() != nil` -- this type never touches Security.framework.
public enum WidgetState: Equatable, Sendable {
    case unpaired
    case empty(headline: String)
    case tasks([TaskItem], counts: MobileSummary.Counts, ageMinutes: Int)
    case stale(Int)
    // Whole-branch review finding: the widget previously had no way to say
    // "the daemon is known to be unreachable" as distinct from "this is just
    // old" -- `RefreshOutcome.offline` never reached the snapshot, so a
    // daemon that had been down the whole time rendered identically to a
    // healthy one for up to 60 minutes. The carried `Int` is minutes since
    // the last snapshot that DID succeed (`Snapshot.fetchedAt`, which
    // `recordRefreshFailure` never touches), matching `.stale`'s shape.
    case unreachable(Int)

    // Past this many minutes, the widget must say the data is stale rather
    // than present old rows as though they were current.
    public static let staleAfterMinutes = 60

    public static func from(snapshot: Snapshot?, paired: Bool, now: Date = Date()) -> WidgetState {
        guard paired else { return .unpaired }
        guard let snapshot else { return .empty(headline: "Nothing synced yet.") }
        let age = snapshot.ageInMinutes(now: now)
        // A known-failed refresh is a stronger, more specific signal than
        // mere staleness -- DESIGN.md's copy rules draw the same distinction
        // between "Can't reach OpenAGI" (an error) and "Last synced 3h ago"
        // (a fact stated with no claim about why).
        if snapshot.lastRefreshFailedAt != nil { return .unreachable(age) }
        if age > staleAfterMinutes { return .stale(age) }
        let visible = snapshot.visibleToday
        if visible.isEmpty { return .empty(headline: snapshot.summary.brief.headline) }
        return .tasks(visible, counts: snapshot.visibleCounts, ageMinutes: age)
    }
}

// What the timeline provider hands to WidgetKit: a render-ready state plus
// the date it was computed for.
public struct TodayEntry: TimelineEntry {
    public let date: Date
    public let state: WidgetState

    public init(date: Date, state: WidgetState) {
        self.date = date
        self.state = state
    }
}
